### ISSUE TITLE
Podman support?

### ISSUE THREAD

**Issue by BeayemX** [mikenye/docker-youtube-dl#54]
Should this image work with Podman?

I tried the following commands: 
```
podman run --rm -i -e PGID=$(id -g) -e PUID=$(id -u) -v "$(pwd)":/workdir:rw mikenye/youtube-dl <url>
podman run --rm -i -v "$(pwd)":/workdir mikenye/youtube-dl <url>
```

but they always result in this error:

```
[download] Unable to open file: [Errno 13] Permission denied:
```

**Comment by dreua** [comment]
I fround a quick workaround:
1. Run an interactive session `podman run --rm -i -e PGID=$(id -g) -e PUID=$(id -u) -v "$(pwd)":/workdir:rw,z -t --entrypoint=bash docker.io/mikenye/youtube-dl`
2. Run `yt-dlp <url>` in there, it works just fine.
3. Exit with Ctrl+d

Note that I added a `,z` to the volume mount because I run with selinux, this is also necessary with docker on selinux systems and should have no effect when running without selinux.

I wonder why this happens or what the default entry point is and what went wrong there. I don't have time to investigate further right now but feel free to ping me if you think this is important.

Tested on Fedora 39 which ships with selinux by default. Should work on any podman installation though.

**Comment by dreua** [comment]
Still the same issue. Any interest to fix this here @mikenye or would I need to fork the whole thing?

**Comment by mikenye** [comment]
I'm happy to accept a PR. I don't have much free time to work on this unfortunately.

**Comment by dreua** [comment]
Thanks, that is good to know :)

**Comment by dogvisor** [comment]
This is my alias for yt-dl which works on Fedora with SELinux:
`yt-dl='uid=$(id -u) gid=$(id -g) podman run --rm -i --uidmap=0:1 --uidmap=$uid:0 -e PGID=$gid -e PUID=$uid --security-opt label=disable -v "$(pwd)":/workdir:rw -v ~/.config/yt-dlp.conf:/etc/yt-dlp.conf:ro ghcr.io/mikenye/docker-youtube-dl:latest'`

In short, since podman using rootless containers maps the UID of the host user to the root user on the container, mounts become owned as root by default. As this container starts as root and then drops privileges (to `dockeruser`) to run the script it is unable to write to the root-owned (within the container) `/workdir`. 

The `--uidmap` juggling in the alias above (re-)remaps the root UID in the container (`0`) to the first (`1`) subuid of the current user on the host (instead of the real UID of the current host user.) Then, the UID of the created `dockeruser` within the container (which we have set to our host UID using `-e PUID`) is mapped to the UID of the host user (i.e. the `0`th offset subuid.) 

Clear as mud! ..but written down here for the benefit of future travellers. 

`--userns=keep-id` would have been the preferred solution, paired with a check in the init script to check if privileges were already dropped by the container runtime.

**Comment by dreua** [comment]
That looks like a very complex work around to me. Is that dropping of privileges a good idea or necessary? I mean it is running in a container after all. (But I'm not an expert on security within containers. It just feels a bit odd to me.) Maybe a better solution would to disable dropping of privileges when run in podman, how would you feel about that? My goal would be to get usability as easy as possible, in my vision a single volume mount for the target directory should suffice.

**Comment by dogvisor** [comment]
Yes it is complex, but that was what I needed to do to get it work with the container as-is within this repository.

The following `Containerfile` works for _my_ needs:
```
FROM docker.io/python:3-alpine
RUN <<EOF
adduser -D yt-dlp
apk update && apk add --quiet --no-interactive ffmpeg
EOF
USER yt-dlp
RUN <<EOF
python3 -m venv /home/yt-dlp/venv
/home/yt-dlp/venv/bin/pip install --no-cache-dir yt-dlp
EOF
COPY yt-dlp.conf /home/yt-dlp/yt-dlp.conf
WORKDIR /data
ENTRYPOINT ["/home/yt-dlp/venv/bin/yt-dlp"
```
..along with the alias:
```
yt-dl='podman run --rm -i -v "$(pwd)":/data -v ~/.config/yt-dlp.conf:/etc/yt-dlp.conf:ro --security-opt label=disable --userns=keep-id yt-dlp'
```
For Docker the equivalent alias is:
```
yt-dl='podman run --rm -i -v "$(pwd)":/data -v ~/.config/yt-dlp.conf:/etc/yt-dlp.conf:ro --security-opt label=disable yt-dlp'
```
And the contents of the containers `/home/yt-dlp/yt-dlp.conf` is just:
```
-P /data
--no-cache-dir
```
I haven't done any significant testing on the full set of `yt-dlp` functionality though, but simple tests work fine:
```
➜  ~ yt-dl https://www.youtube.com/watch\?v\=dQw4w9WgXcQ
[youtube] Extracting URL: https://www.youtube.com/watch?v=dQw4w9WgXcQ
[youtube] dQw4w9WgXcQ: Downloading webpage
[youtube] dQw4w9WgXcQ: Downloading tv client config
[youtube] dQw4w9WgXcQ: Downloading player 7d1d50a6
[youtube] dQw4w9WgXcQ: Downloading tv player API JSON
[youtube] dQw4w9WgXcQ: Downloading ios player API JSON
[youtube] dQw4w9WgXcQ: Downloading m3u8 information
[info] dQw4w9WgXcQ: Downloading 1 format(s): 616+251
[info] There are no subtitles for the requested languages
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 39
[download] Destination: /data/Rick Astley - Never Gonna Give You Up (Official Music Video) [dQw4w9WgXcQ].f616.mp4
[download] 100% of   94.86MiB in 00:00:14 at 6.58MiB/s
[download] Destination: /data/Rick Astley - Never Gonna Give You Up (Official Music Video) [dQw4w9WgXcQ].f251.webm
[download] 100% of    3.28MiB in 00:00:00 at 12.69MiB/s
[Merger] Merging formats into "/data/Rick Astley - Never Gonna Give You Up (Official Music Video) [dQw4w9WgXcQ].webm"
Deleting original file /data/Rick Astley - Never Gonna Give You Up (Official Music Video) [dQw4w9WgXcQ].f616.mp4 (pass -k to keep)
Deleting original file /data/Rick Astley - Never Gonna Give You Up (Official Music Video) [dQw4w9WgXcQ].f251.webm (pass -k to keep)
➜  ~ ls -lh Rick\ Astley\ -\ Never\ Gonna\ Give\ You\ Up\ \(Official\ Music\ Video\)\ \[dQw4w9WgXcQ\].webm
-rw-r--r--. 1 dogvisor dogvisor 99M May 30  2024 'Rick Astley - Never Gonna Give You Up (Official Music Video) [dQw4w9WgXcQ].webm'
```

**Comment by dogvisor** [comment]
The complexities in the existing container (with juggling UIDs etc) are due to it creating a user, `chmod`'ing the target directory and dropping privileges on each run in-container.
